Add "Web Components v0 deprecation" article. (#6890)
google/WebFundamentals: src/content/en/updates/2019/01/webcomponents-v0-deprecations.md

[Chromium Bug](https://crbug.com/766694)

## Migration support
If your site and/or library is using this features, you should upgrade them to
the v1 APIs.
Use `$ chrome --disable-blink-features=ShadowDOMV0,CustomElementsV0,HTMLImports`
on command line to launch Chrome disabled v0 features for testing if your site works without those APIs.

### Polyfill

Now webcomponents polyfill supports v0 features on the all-v0-disabled Chrome. A web site using the polyfill should continue to work.
This is also another option for the web authors.
You can install the polyfill on the site using v0 features directly to keep
it working.
There are also other alternative polyfills for each feature (e.g.
[AshleyScirra’s HTML Imports polyfill](https://github.com/AshleyScirra/html-imports-polyfill)).

### Postpone removal via origin trial

For web authors who need more time to handle this removal, you can re-enable
these features with an [origin trial](https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/developer-guide.md).
After registering for the trial, you will receive a token that allows your site to use the three deprecated features for up to a year,
until Chrome 81.

> Chrome 81 is the planned milestone when we remove implementation code entirely so

Author: ```suggestion
```

> the trial token might last earlier.

Author: ```suggestion
```

> 

Author: ```suggestion
```

> The trial name is `WebComponentsV0`. With that token, you can use Shadow DOM v0,

Author: ```suggestion
```

> Custom Elements v0, and HTML Imports APIs. Trial registration will be open

Reviewer: My earlier [comment](https://github.com/google/WebFundamentals/pull/6890#discussion_r247648880) could have been more clear. I suggested new content, but that intended to replace existing paragraphs, not as an addition. I see the suggested content was added, so now need to remove the two paragraphs, starting with "Chrome 81 is the .." and ending with "... registration will be open" (lines 65 - 69 in the current version of the file).

Author: Done

> Custom Elements v0, and HTML Imports APIs. Trial registration will be open

Author: ```suggestion
```

 and [HTML Imports](https://www.html5rocks.com/en/tutorials/webcomponents/imports/)
  in Mar, 2019 at Chrome 73.

[Intent to Deprecate](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/h-JwMiPUnuU/sl79aLoLBQAJ)

## Motivation

Shadow DOM v0, Custom Elements v0, and HTML Imports (hereinafter, called v0 features)
 were launched in 2014, but
they did not get adopted by other browser engines. Instead, Shadow DOM v1,
Custom Elements v1, and ES modules are widely adopted by various browser engines
today. Chrome has shipped Shadow DOM v1 / Custom Elements v1 in 2016 and ES
modules in 2017.

- Shadow DOM v0: [Chromestatus Tracker](https://www.chromestatus.com/feature/4507242028072960) &#124;
 [Chromium Bug](https://crbug.com/671907)
- Custom Elements v0: [Chromestatus Tracker](https://www.chromestatus.com/feature/4642138092470272)  &#124;
 [Chromium Bug](https://crbug.com/660759)
- HTML Imports: [Chromestatus Tracker](https://www.chromestatus.com/feature/4507242028072960)  &#124;
[Chromium Bug](https://crbug.com/766694)

### Why are all APIs being removed at once?

Chrome usually removes features one by one, but why are these features being removed at the same time?

> Since launched, we found that the v0 APIsy are often used together.

Author: Done. I understand the circumstance that we don't want to say "because of poly*" no matter that's good or bad.

The [webcomponentsjs polyfill](https://github.com/WebComponents/webcomponentsjs)
 was originally designed to simulate the v0 features on other
browsers and not to do same on Chrome being removed each feature one by one.
For safer/lesser troublesome migration off from the features, we've decided to
remove them at once.

## Migration support
If your site and/or library is using this features, you should upgrade them.
Use `$ chrome --disable-blink-features=ShadowDOMV0,CustomElementsV0,HTMLImports`
on command line to launch Chrome disabled v0 features for testing if your site works without those APIs.

### Polyfill

Now webcomponents polyfill supports v0 features on the all-v0-disabled Chrome. A web site using the polyfill should continue to work.
This is also another option for the web authors.
You can install the polyfill on the site using v0 features directly to keep
it working.
There are also other alternative polyfills for each feature (e.g.
[AshleyScirra’s HTML Imports polyfill](https://github.com/AshleyScirra/html-imports-polyfill)).

### Reverse Origin Trial

> For web authors that don't much time to Mar, 2019 at Chrome 73,

Reviewer 2: or maybe "for up to an additional year", if the tokens aren't going to last for the entire year. Do we know if that has been decided yet?

> For web authors that don't much time to Mar, 2019 at Chrome 73,

Author: Shorter is better, of course. Chrome M81 is planned to be the first Chrome that has no capability to access the APIs .

- Shadow DOM v0: [Chromestatus Tracker](https://www.chromestatus.com/feature/4507242028072960) &#124;
 [Chromium Bug](https://crbug.com/671907)
- Custom Elements v0: [Chromestatus Tracker](https://www.chromestatus.com/feature/4642138092470272)  &#124;
 [Chromium Bug](https://crbug.com/660759)
- HTML Imports: [Chromestatus Tracker](https://www.chromestatus.com/feature/4507242028072960)  &#124;
[Chromium Bug](https://crbug.com/766694)

## Migration support
If your site and/or library is using this features, you should upgrade them to
the v1 APIs.
Use `$ chrome --disable-blink-features=ShadowDOMV0,CustomElementsV0,HTMLImports`
on command line to launch Chrome disabled v0 features for testing if your site works without those APIs.

### Polyfill

Now webcomponents polyfill supports v0 features on the all-v0-disabled Chrome. A web site using the polyfill should continue to work.
This is also another option for the web authors.
You can install the polyfill on the site using v0 features directly to keep
it working.
There are also other alternative polyfills for each feature (e.g.
[AshleyScirra’s HTML Imports polyfill](https://github.com/AshleyScirra/html-imports-polyfill)).

> ### Reverse Origin Trial

Reviewer: ```suggestion
### Postpone removal via origin trial
```
Suggesting a different heading:
- I'd prefer not to introduce a term like "reverse origin trial" that could be difficult for developers to understand.
- Minor nit is that "origin trials" is generally not capitalized. It is not a branded name.

For web authors who need more time to handle this removal, you can re-enable

> these features with an [Origin Trial](https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/developer-guide.md).

Reviewer: ```suggestion
these features with an [origin trial](https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/developer-guide.md).
```
Lower case for "origin trial"

> With a token generated for the trial, your site can continue to use them for

Reviewer: ```suggestion
After registering for the trial, you will receive a token that allows your site to use the three deprecated features for up to a year,
```

> an additional year, until Chrome 81.

Reviewer: ```suggestion
until Chrome 81.
```

Chrome 81 is the planned milestone when we remove implementation code entirely so
the trial token might last earlier.

The trial name is `WebComponentsV0`. With that token, you can use Shadow DOM v0,
Custom Elements v0, and HTML Imports APIs. Trial registration will be open

> after Chrome 73 beta release."

Reviewer: I suggest simplifying the preceding two paragraphs to something like:

```suggestion
We plan to remove the features permanently in the Chrome 81 release, although the trial tokens may expire before then.

Soon after the Chrome 73 beta release, you can register for the trial at [Web Components V0](https://developers.chrome.com/origintrials/#/view_trial/2431943798780067841).
```
- The text above (with suggested edits) already explains that the trial will re-enable all 3 features, so probably don't need to repeat that here.
- I've added the link to the specific trial. This link will not work until the trial is made public, but it will redirect to the list of trials in the meantime.
